### PR TITLE
Add add_to_cleanup fixture

### PR DESCRIPTION
**Pull request by mdellweg**
This fixture will collect references to pulp objects to be deleted after
a test. It will attempt deletion in reversed order and await all tasks
in parallel.

### CODE REVIEW

**Comment by mdellweg** [comment]
Rationale: In pulp container we create Distributions and namespaces by push operations, so we cannot use `gen_object_with_cleanup` there. Also we do not have `unittest.TestCase.teardown()` available anymore.

File: pulp_smash/pulp3/pytest_plugin/__init__.py [hunk 1]
```diff
@@ -403,7 +403,9 @@ def __exit__(self, exc_type, exc_value, traceback):
 
 
 @pytest.fixture
-def delete_orphans_pre():
+def delete_orphans_pre(request):
+    if request.node.get_closest_marker("parallel") is not None:
+        raise pytest.UsageError("This test is no suitable to be marked parallel.")
```
Comment (bmbouter) [inline pulp_smash/pulp3/pytest_plugin/__init__.py:408]: Nice!

File: pulp_smash/pulp3/pytest_plugin/__init__.py [hunk 2]
```diff
@@ -506,13 +508,35 @@ def ssl_ctx_req_client_auth(
...
 
 @pytest.fixture
-def gen_object_with_cleanup():
-    obj_hrefs = []
+def add_to_cleanup():
+    """Fixture to allow pulp objects to be deleted after the test."""
+    obj_refs = []
 
+    def _add_to_cleanup(api_client, pulp_href):
+        obj_refs.insert(0, (api_client, pulp_href))
```
Comment (bmbouter) [inline pulp_smash/pulp3/pytest_plugin/__init__.py:516]: Python lists are not optimized to insert at the beginning, but the collections.deque is optimized to insert from both ends. I think we should use that.

**Review by bmbouter — CHANGES_REQUESTED** [review]
I'm requesting one data structure change please.

File: pulp_smash/pulp3/pytest_plugin/__init__.py (hunk 2 above)
Comment (mdellweg) [reply pulp_smash/pulp3/pytest_plugin/__init__.py:516]: What about calling `obj_refs.reverse()` before cleanup? Or using a reverse iterator (does python have such a thing?)?

File: pulp_smash/pulp3/pytest_plugin/__init__.py (hunk 1 above)
Comment (mdellweg) [reply pulp_smash/pulp3/pytest_plugin/__init__.py:408]:
```suggestion
        raise pytest.UsageError("This test is not suitable to be marked parallel.")
```

File: pulp_smash/pulp3/pytest_plugin/__init__.py (hunk 2 above)
Comment (bmbouter) [reply pulp_smash/pulp3/pytest_plugin/__init__.py:516]: Then we have to reverse them? I think this is what `collections.deque` was made for actually so I think we should use it.

File: pulp_smash/pulp3/pytest_plugin/__init__.py [hunk 3]
```diff
@@ -525,13 +527,35 @@ def ssl_ctx_req_client_auth(
...
+    """Fixture to allow pulp objects to be deleted after the test."""
+    obj_refs = []
 
+    def _add_to_cleanup(api_client, pulp_href):
+        obj_refs.append((api_client, pulp_href))
+
+    yield _add_to_cleanup
+
+    delete_task_hrefs = []
+    for api_client, pulp_href in reversed(obj_refs):
```
Comment (bmbouter) [inline pulp_smash/pulp3/pytest_plugin/__init__.py:540]: Can you put a little commend on why processing in reverse order is important?

**Review by bmbouter — APPROVED** [review]
Thanks!

**Pull request merged**